### PR TITLE
Reduce unnecessary Disk IO

### LINKED ISSUE #1958: [FR] Reduce unnecessary disk IO
# Feature Request

## Context
```check-for-changes.sh``` script is constantly (every second) hashing and writing unnecessary to host filesystem.

### Is your Feature Request related to a Problem?
Reduce unnecessary disk IO.

### Describe the Solution you'd like
Wait and hash only if change was detected.

### Are you going to implement it?
I have just a quick hack but it depends on package ```inotify-tools```.
I have added in ```check-for-changes.sh``` before line 46:
```
  inotifywait -r \
          /tmp/docker-mailserver \
          /etc/letsencrypt/acme.json \
          "/etc/letsencrypt/live"
```

But this snippet is not heavily. Have just add it to my container.

### PR DESCRIPTION
**Pull request by aendeavor**
# Description

See #1958

@nukedupe please double-check this PR too.

Fixes #1958

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by aendeavor** [comment]
@nukedupe Please have a look at CI:

```
not ok 111 can detect changes
# (from function `assert_output' in file test/test_helper/bats-assert/src/assert.bash, line 231,
#  in test file test/mail_ssl_letsencrypt.bats, line 116)
#   `assert_output --partial "postfix: stopped"' failed
# 
# -- output does not contain substring --
# substring (1 lines):
#   postfix: stopped
# output (4 lines):
#   [[ TASKS ]]  2021-05-12 10:07:45  Start check-for-changes script.
#   [[  INF  ]]  2021-05-12 10:07:45  Using postmaster address postmaster@my-domain.com
#   Setting up watches.  Beware: since -r was given, this may take a while!
#   Watches established.
# --
# 
```

Your changes do not actually achieve what you intended, or CI needs adjustments. The former is way more likely.

**Comment by aendeavor** [comment]
@docker-mailserver/maintainers This is ready for review :)

File: target/scripts/check-for-changes.sh [hunk 1]
```diff
@@ -41,10 +41,10 @@ sleep 10
 
 while true
 do
-  inotifywait -r \
-    /tmp/docker-mailserver \
-    /etc/letsencrypt/acme.json \
-    /etc/letsencrypt/live
+  inotifywait -qq -r \
+    -e modify -e create -e delete -e moved \
```
Comment (casperklein) [inline target/scripts/check-for-changes.sh:45]: There is no "move" event: https://man.cx/inotifywait#heading7

Comment (aendeavor) [reply target/scripts/check-for-changes.sh:45]: You mean, there is no "move**d**" , event? I will correct this, the event is called "move", without the "d". Is that what you mean?

Comment (aendeavor) [reply target/scripts/check-for-changes.sh]: Interestingly, tests fail again. Oh nice.

**Comment by nukedupe** [comment]
Test can't detect change because there is no event (modify,create, delete or move). cp is executed before inotifywait runs and wait for a event.
Wrong order in ```test/mail_ssl_letsencrypt.bats``` Maybe something like ?
```
@test "can detect changes" {
  run docker exec mail_lets_acme_json /bin/bash -c "supervisorctl tail changedetector"
  sleep 15
  cp "$(private_config_path mail_lets_acme_json)/letsencrypt/acme-changed.json" "$(private_config_path mail_lets_acme_json)/acme.json"
  assert_output --partial "postfix: stopped"
  assert_output --partial "postfix: started"
  assert_output --partial "Change detected"
```

File: target/scripts/check-for-changes.sh (hunk 1 above)
Comment (nukedupe) [reply target/scripts/check-for-changes.sh]: My mistake, I haven't tested letsencrypt stuff. Seems like there is a problem with inotifywait on volume mounts.
`echo "" >> /tmp/docker-mailserver/postfix-aliases.cf` works
`echo "" >> /etc/letsencrypt/acme.json` works not

Does something speak against using `sha512sum -c --status /tmp/docker-mailserver-config-chksum` and check return value and not write to temp file and do `cmp --silent ...` ?

Comment (aendeavor) [reply target/scripts/check-for-changes.sh]: > My mistake, I haven't tested letsencrypt stuff. Seems like there is a problem with inotifywait on volume mounts.
> 
> `echo "" >> /tmp/docker-mailserver/postfix-aliases.cf` works
> 
> `echo "" >> /etc/letsencrypt/acme.json` works not
> 
> 
> 
> Does something speak against using `sha512sum -c --status /tmp/docker-mailserver-config-chksum` and check return value and not write to temp file and do `cmp --silent ...` ?

No, nothing speaks against it. Can you open a PR yourself? I would then close this PR.

Comment (nukedupe) [reply target/scripts/check-for-changes.sh]: Here is a better solution. But my PR is not in this PR... have to learn git... Can you fix it?

Comment (aendeavor) [reply target/scripts/check-for-changes.sh]: Sorry, very low on time lately. Learning GIT is always beneficial, trust me :) I will close this - and if you have a working solution, create a pull request on GitHub.

PS: Have a look at our `CONTRIBUTING.md`, which explains how to do a pull request and add code on detail :)

**Pull request closed without merging**